docs/man/man3/aDbSetup.3
.TH "aDbSetup" 3 "Fri Mar 4 2022" "openPilotLog" \" -*- nroff -*-
.ad l
.nh
.SH NAME
aDbSetup \- The \fBaDbSetup\fP namespace is responsible for the inital setup of the database when the application is first launched\&. It creates the database in the specified default location and creates all required tables and views\&.  

.SH SYNOPSIS
.br
.PP
.SS "Functions"

.in +1c
.ti -1c
.RI "bool \fBcreateDatabase\fP ()"
.br
.RI "createDatabase runs a number of CREATE queries that create the database tables and columns\&. "
.ti -1c
.RI "bool \fBcommitData\fP (const QJsonArray &json_arr, const QString &table_name)"
.br
.RI "commitData commits the data read from a JSON file into a table in the database\&. "
.ti -1c
.RI "bool \fBimportTemplateData\fP (bool use_local_ressources)"
.br
.RI "importTemplateData fills an empty database with the template data (Aircraft, Airports, currencies, changelog) as read from the JSON templates\&. "
.ti -1c
.RI "bool \fBresetUserData\fP ()"
.br
.RI "Empties all user-generated content in the database\&. "
.in -1c
.SS "Variables"

.in +1c
.ti -1c
.RI "const auto \fBTEMPLATE_URL\fP = QStringLiteral('https://raw\&.githubusercontent\&.com/fiffty\-50/openpilotlog/develop/assets/database/templates/')"
.br
.ti -1c
.RI "const auto \fBCREATE_TABLE_PILOTS\fP"
.br
.ti -1c
.RI "const auto \fBCREATE_TABLE_TAILS\fP"
.br
.ti -1c
.RI "const auto \fBCREATE_TABLE_FLIGHTS\fP"
.br
.ti -1c
.RI "const auto \fBCREATE_TABLE_AIRPORTS\fP"
.br
.ti -1c
.RI "const auto \fBCREATE_TABLE_AIRCRAFT\fP"
.br
.ti -1c
.RI "const auto \fBCREATE_TABLE_CHANGELOG\fP"
.br
.ti -1c
.RI "const auto \fBCREATE_TABLE_CURRENCIES\fP"
.br
.ti -1c
.RI "const auto \fBCREATE_VIEW_DEFAULT\fP"
.br
.ti -1c
.RI "const auto \fBCREATE_VIEW_EASA\fP"
.br
.ti -1c
.RI "const auto \fBCREATE_VIEW_TAILS\fP"
.br
.ti -1c
.RI "const auto \fBCREATE_VIEW_PILOTS\fP"
.br
.ti -1c
.RI "const auto \fBCREATE_VIEW_QCOMPLETER\fP"
.br
.ti -1c
.RI "const auto \fBCREATE_VIEW_TOTALS\fP"
.br
.ti -1c
.RI "const QList< QLatin1String > \fBDATABASE_TABLES\fP"
.br
.ti -1c
.RI "const QList< QLatin1String > \fBDATABASE_VIEWS\fP"
.br
.ti -1c
.RI "const QList< QLatin1String > \fBUSER_TABLES\fP"
.br
.ti -1c
.RI "const QList< QLatin1String > \fBTEMPLATE_TABLES\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
The \fBaDbSetup\fP namespace is responsible for the inital setup of the database when the application is first launched\&. It creates the database in the specified default location and creates all required tables and views\&. 
.SH "Function Documentation"
.PP 
.SS "bool aDbSetup::createDatabase ()"

.PP
createDatabase runs a number of CREATE queries that create the database tables and columns\&. 
.PP
\fBReturns\fP
.RS 4

.RE
.PP

.SS "bool aDbSetup::importTemplateData (bool use_local_ressources)"

.PP
importTemplateData fills an empty database with the template data (Aircraft, Airports, currencies, changelog) as read from the JSON templates\&. 
.PP
\fBParameters\fP
.RS 4
\fIuse_local_ressources\fP determines whether the included ressource files or a previously downloaded file should be used\&. 
.RE
.PP
\fBReturns\fP
.RS 4
.RE
.PP

.SS "bool aDbSetup::resetUserData ()"

.PP
Empties all user-generated content in the database\&. 
.PP
\fBReturns\fP
.RS 4
true on success 
.RE
.PP

.SH "Variable Documentation"
.PP 
.SS "const auto aDbSetup::CREATE_TABLE_AIRCRAFT"
\fBInitial value:\fP
.PP
.nf
= QLatin1String("CREATE TABLE aircraft ("
            " aircraft_id   INTEGER NOT NULL,"
            " make          TEXT,"
            " model         TEXT,"
            " variant       TEXT,"
            " name          TEXT,"
            " iata          TEXT,"
            " icao          TEXT,"
            " multipilot    INTEGER,"
            " multiengine   INTEGER,"
            " engineType    INTEGER,"
            " weightClass   INTEGER,"
            " PRIMARY KEY(aircraft_id AUTOINCREMENT)"
            ")")
.fi
.SS "const auto aDbSetup::CREATE_TABLE_AIRPORTS"
\fBInitial value:\fP
.PP
.nf
= QLatin1String("CREATE TABLE airports ( "
            " airport_id     INTEGER NOT NULL, "
            " icao           TEXT NOT NULL, "
            " iata           TEXT, "
            " name           TEXT, "
            " lat            REAL, "
            " long           REAL, "
            " country        TEXT, "
            " alt            INTEGER, "
            " utcoffset      INTEGER, "
            " tzolson        TEXT, "
            " PRIMARY KEY(airport_id AUTOINCREMENT) "
            ")")
.fi
.SS "const auto aDbSetup::CREATE_TABLE_CHANGELOG"
\fBInitial value:\fP
.PP
.nf
= QLatin1String("CREATE TABLE changelog ( "
            " revision   INTEGER NOT NULL, "
            " comment    TEXT, "
            " date       NUMERIC, "
            " PRIMARY KEY(revision) "
            ")")
.fi
.SS "const auto aDbSetup::CREATE_TABLE_CURRENCIES"
\fBInitial value:\fP
.PP
.nf
= QLatin1String("CREATE TABLE currencies ( "
            " currency_id   INTEGER PRIMARY KEY AUTOINCREMENT, "
            " description   TEXT, "
            " expiryDate     NUMERIC "
            ")"
            )
.fi
.SS "const auto aDbSetup::CREATE_TABLE_PILOTS"
\fBInitial value:\fP
.PP
.nf
= QLatin1String("CREATE TABLE pilots ( "
            " pilot_id       INTEGER NOT NULL, "
            " lastname       TEXT    NOT NULL, "
            " firstname      TEXT, "
            " alias          TEXT, "
            " company        TEXT, "
            " employeeid     TEXT, "
            " phone          TEXT, "
            " email          TEXT, "
            " PRIMARY KEY(pilot_id AUTOINCREMENT)"
            ")")
.fi
.SS "const auto aDbSetup::CREATE_TABLE_TAILS"
\fBInitial value:\fP
.PP
.nf
= QLatin1String("CREATE TABLE tails ("
            " tail_id        INTEGER NOT NULL,"
            " registration   TEXT NOT NULL,"
            " company        TEXT,"
            " make           TEXT,"
            " model          TEXT,"
            " variant        TEXT,"
            " multipilot     INTEGER,"
            " multiengine    INTEGER,"
            " engineType     INTEGER,"
            " weightClass    INTEGER,"
            " PRIMARY KEY(tail_id AUTOINCREMENT)"
            ")")
.fi
.SS "const auto aDbSetup::CREATE_VIEW_DEFAULT"
\fBInitial value:\fP
.PP
.nf
= QLatin1String("CREATE VIEW viewDefault AS "
        " SELECT flight_id, doft as 'Date', "
        " dept AS 'Dept', "
        " printf('%02d',(tofb/60))||':'||printf('%02d',(tofb%60)) AS 'Time', "
        " dest AS 'Dest', printf('%02d',(tonb/60))||':'||printf('%02d',(tonb%60)) AS 'Time ', "
        " printf('%02d',(tblk/60))||':'||printf('%02d',(tblk%60)) AS 'Total', "
        " CASE "
        " WHEN pilot_id = 1 THEN alias "
        " ELSE lastname||', '||substr(firstname, 1, 1)||'\&.' "
        " END "
        " AS 'Name PIC', "
        " CASE "
        " WHEN variant IS NOT NULL THEN make||' '||model||'-'||variant "
        " ELSE make||' '||model "
        " END "
        " AS 'Type', "
        " registration AS 'Registration', "
        " FlightNumber AS 'Flight #', "
        " remarks AS 'Remarks'"
        " FROM flights "
        " INNER JOIN pilots on flights\&.pic = pilots\&.pilot_id "
        " INNER JOIN tails on flights\&.acft = tails\&.tail_id "
        " ORDER BY date DESC ")
.fi
.SS "const auto aDbSetup::CREATE_VIEW_PILOTS"
\fBInitial value:\fP
.PP
.nf
= QLatin1String("CREATE VIEW viewPilots AS "
        " SELECT "
        " pilot_id AS 'ID', "
        " lastname AS 'Last Name', "
        " firstname AS 'First Name', "
        " company AS 'Company' "
        " FROM pilots")
.fi
.SS "const auto aDbSetup::CREATE_VIEW_QCOMPLETER"
\fBInitial value:\fP
.PP
.nf
= QLatin1String("CREATE VIEW viewQCompleter AS "
        " SELECT airport_id, icao, iata, tail_id, registration, pilot_id, "
        " lastname||', '||firstname AS 'pilot_name', alias "
        " FROM airports "
        " LEFT JOIN tails ON airports\&.airport_id = tails\&.tail_id "
        " LEFT JOIN pilots ON airports\&.airport_id = pilots\&.pilot_id")
.fi
.SS "const auto aDbSetup::CREATE_VIEW_TAILS"
\fBInitial value:\fP
.PP
.nf
= QLatin1String("CREATE VIEW viewTails AS "
        " SELECT "
        " tail_id AS 'ID', "
        " registration AS 'Registration', "
        " make||' '||model AS 'Type', "
        " company AS 'Company' "
        " FROM tails WHERE model IS NOT NULL AND variant IS NULL "
        " UNION "
        " SELECT "
        " tail_id AS 'ID', "
        " registration AS 'Registration', "
        " make||' '||model||'-'||variant AS 'Type', "
        " company AS 'Company' "
        " FROM tails WHERE variant IS NOT NULL")
.fi
.SS "const auto aDbSetup::CREATE_VIEW_TOTALS"
\fBInitial value:\fP
.PP
.nf
= QLatin1String("CREATE VIEW viewTotals AS "
        " SELECT "
        " printf(\"%02d\",CAST(SUM(tblk) AS INT)/60)||\":\"||printf(\"%02d\",CAST(SUM(tblk) AS INT)%60) AS \"TOTAL\", "
        " printf(\"%02d\",CAST(SUM(tSPSE) AS INT)/60)||\":\"||printf(\"%02d\",CAST(SUM(tSPSE) AS INT)%60) AS \"SP SE\", "
        " printf(\"%02d\",CAST(SUM(tSPME) AS INT)/60)||\":\"||printf(\"%02d\",CAST(SUM(tSPME) AS INT)%60) AS \"SP ME\", "
        " printf(\"%02d\",CAST(SUM(tNIGHT) AS INT)/60)||\":\"||printf(\"%02d\",CAST(SUM(tNIGHT) AS INT)%60) AS \"NIGHT\", "
        " printf(\"%02d\",CAST(SUM(tIFR) AS INT)/60)||\":\"||printf(\"%02d\",CAST(SUM(tIFR) AS INT)%60) AS \"IFR\", "
        " printf(\"%02d\",CAST(SUM(tPIC) AS INT)/60)||\":\"||printf(\"%02d\",CAST(SUM(tPIC) AS INT)%60) AS \"PIC\", "
        " printf(\"%02d\",CAST(SUM(tPICUS) AS INT)/60)||\":\"||printf(\"%02d\",CAST(SUM(tPICUS) AS INT)%60) AS \"PICUS\", "
        " printf(\"%02d\",CAST(SUM(tSIC) AS INT)/60)||\":\"||printf(\"%02d\",CAST(SUM(tSIC) AS INT)%60) AS \"SIC\", "
        " printf(\"%02d\",CAST(SUM(tDual) AS INT)/60)||\":\"||printf(\"%02d\",CAST(SUM(tDual) AS INT)%60) AS \"DUAL\", "
        " printf(\"%02d\",CAST(SUM(tFI) AS INT)/60)||\":\"||printf(\"%02d\",CAST(SUM(tFI) AS INT)%60) AS \"INSTRUCTOR\", "
        " printf(\"%02d\",CAST(SUM(tSIM) AS INT)/60)||\":\"||printf(\"%02d\",CAST(SUM(tSIM) AS INT)%60) AS \"SIMULATOR\", "
        " printf(\"%02d\",CAST(SUM(tMP) AS INT)/60)||\":\"||printf(\"%02d\",CAST(SUM(tMP) AS INT)%60) AS \"MultPilot\", "
        " CAST(SUM(toDay) AS INT) AS \"TO Day\", CAST(SUM(toNight) AS INT) AS \"TO Night\", "
        " CAST(SUM(ldgDay) AS INT) AS \"LDG Day\", CAST(SUM(ldgNight) AS INT) AS \"LDG Night\" "
        " FROM flights")
.fi
.SS "const QList<QLatin1String> aDbSetup::DATABASE_TABLES"
\fBInitial value:\fP
.PP
.nf
= {
    CREATE_TABLE_PILOTS,
    CREATE_TABLE_TAILS,
    CREATE_TABLE_FLIGHTS,
    CREATE_TABLE_AIRCRAFT,
    CREATE_TABLE_AIRPORTS,
    CREATE_TABLE_CURRENCIES,
    CREATE_TABLE_CHANGELOG
}
.fi
.SS "const QList<QLatin1String> aDbSetup::DATABASE_VIEWS"
\fBInitial value:\fP
.PP
.nf
= {
    CREATE_VIEW_DEFAULT,
    CREATE_VIEW_EASA,
    CREATE_VIEW_TAILS,
    CREATE_VIEW_PILOTS,
    CREATE_VIEW_TOTALS,
    CREATE_VIEW_QCOMPLETER,
}
.fi
.SS "const QList<QLatin1String> aDbSetup::TEMPLATE_TABLES"
\fBInitial value:\fP
.PP
.nf
= {
    QLatin1String("aircraft"),
    QLatin1String("airports"),
    QLatin1String("currencies"),
    QLatin1String("changelog")
}
.fi
.SS "const QList<QLatin1String> aDbSetup::USER_TABLES"
\fBInitial value:\fP
.PP
.nf
= {
    QLatin1String("flights"),
    QLatin1String("pilots"),
    QLatin1String("tails")
}
.fi
.SH "Author"
.PP 
Generated automatically by Doxygen for openPilotLog from the source code\&.


CHANGELOG.md
# openpilotlog changelog


New folder and organizational structure
note: Database and settings stored in "data" folder in application directory

All database functionality now in one class

Using QSettings class instead of custom solution (storing settings in database)
settings table in database now obsolete

New class completionLists, provides lists for QCompleter

New class aircraft

Rework of New Aircraft Dialog (WIP)


src/database/adbsetup.cpp
/*
 *openPilotLog - A FOSS Pilot Logbook Application
 *Copyright (C) 2020-2021 Felix Turowsky
 *
 *This program is free software: you can redistribute it and/or modify
 *it under the terms of the GNU General Public License as published by
 *the Free Software Foundation, either version 3 of the License, or
 *(at your option) any later version.
 *
 *This program is distributed in the hope that it will be useful,
 *but WITHOUT ANY WARRANTY; without even the implied warranty of
 *MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *GNU General Public License for more details.
 *
 *You should have received a copy of the GNU General Public License
 *along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
#include "adbsetup.h"
#include "src/opl.h"
#include "src/database/adatabase.h"
#include "src/functions/alog.h"
#include "src/classes/ajson.h"

namespace aDbSetup {

// const auto TEMPLATE_URL = QStringLiteral("https://raw.githubusercontent.com/fiffty-50/openpilotlog/tree/main/assets/database/templates/");
const static auto TEMPLATE_URL = QStringLiteral("https://raw.githubusercontent.com/fiffty-50/openpilotlog/develop/assets/database/templates/");

const static QStringList USER_TABLES = {
    QStringLiteral("flights"),
    QStringLiteral("pilots"),
    QStringLiteral("tails")
};
const static QStringList TEMPLATE_TABLES= {
    QStringLiteral("aircraft"),
    QStringLiteral("airports"),
    QStringLiteral("currencies"),
    QStringLiteral("changelog")
};

bool createDatabase()
{
    // Read Database layout from sql file
    QFile f(OPL::Assets::DATABASE_SCHEMA);
    f.open(QIODevice::ReadOnly);
    QByteArray filedata = f.readAll();
    // create individual queries for each table/view
    auto list = filedata.split(';');

    // Commit
    QSqlQuery q;
    QVector<QSqlError> errors;
    // Create Tables
    for (const auto &query_string : list) {
        q.prepare(query_string);
        if (!q.exec()) {
            errors.append(q.lastError());
            LOG << "Unable to execute query: ";
            LOG << q.lastQuery();
            LOG << q.lastError();
        }
    }

    aDB->updateLayout();

    if (errors.isEmpty()) {
        LOG << "Database succesfully created.";
        return true;
    } else {
        LOG << "Database creation has failed. The following error(s) have ocurred: ";
        for (const auto &error : qAsConst(errors)) {
            LOG << error.type() << error.text();
        }
        return false;
    }
}

bool commitData(const QJsonArray &json_arr, const QString &table_name)
{
    //aDB->updateLayout(); // needed?
    QSqlQuery q;

    // create insert statement
    QString statement = QLatin1String("INSERT INTO ") + table_name + QLatin1String(" (");
    QString placeholder = QStringLiteral(") VALUES (");
    for (const auto &column_name : aDB->getTableColumns(table_name)) {
        statement += column_name + ',';
        placeholder.append(QLatin1Char(':') + column_name + QLatin1Char(','));
    }

    statement.chop(1);
    placeholder.chop(1);
    placeholder.append(')');
    statement.append(placeholder);

    q.prepare(QStringLiteral("BEGIN EXCLUSIVE TRANSACTION"));
    q.exec();
    //DEB << statement;
    for (const auto &entry : json_arr) {
        q.prepare(statement);

        auto object = entry.toObject();
        const auto keys = object.keys();
        for (const auto &key : keys){
            object.value(key).isNull() ? q.bindValue(key, QVariant(QVariant::String)) :
                                         q.bindValue(QLatin1Char(':') + key, object.value(key).toVariant());
        }

        q.exec();
    }

    q.prepare(QStringLiteral("COMMIT"));
    if (q.exec())
        return true;
    else
        return false;
}

bool importTemplateData(bool use_local_ressources)
{
    //QSqlQuery q;
    // reset template tables
    const auto table_names = aDB->getTemplateTableNames();
    for (const auto& table_name : table_names) {

        //clear table
        //q.prepare(QLatin1String("DELETE FROM ") + table_name);
        //if (!q.exec()) {
        //    DEB << "Error: " << q.lastError().text();
        //    return false;
        //}

        //Prepare data
        QJsonArray data_to_commit;
        QString error_message("Error importing data ");

        if (use_local_ressources) {
            data_to_commit = AJson::readFileToDoc(QLatin1String(":database/templates/")
                                      + table_name + QLatin1String(".json")).array();
            error_message.append(QLatin1String(" (ressource) "));
        } else {
            data_to_commit = AJson::readFileToDoc(AStandardPaths::directory(
                                          AStandardPaths::Templates).absoluteFilePath(
                                          table_name + QLatin1String(".json"))).array();
            error_message.append(QLatin1String(" (downloaded) "));
        }

        // commit Data from Array
        if (!commitData(data_to_commit, table_name)) {
            LOG << error_message;
            return false;
        }
    } // for table_name
    return true;
}

bool resetUserData()
{
    QSqlQuery query;

    // clear user tables
    for (const auto& table : USER_TABLES) {
        query.prepare("DELETE FROM " + table);
        if (!query.exec()) {
            DEB << "Error: " << query.lastError().text();
        }
    }
    return true;
}

} // namespace aDbSetup
